pkg/auth/dex.go

```diff
@@ -0,0 +1,247 @@
+	ClientSecret string
+	// The Dex redirect callback. Needs to match the dex staticClients.redirectURIs helm configuration.
+	RedirectURI string
+	// The available scopes.
+	Scopes []string
+	// The http client used.
+	Client *http.Client
+}
+
+const (
+	// Dex service internal URL. Used to connect to dex internally in the cluster.
+	dexServiceURL = "http://kuberpult-dex-service:5556"
+	// Dex issuer path. Needs to be match the dex issuer helm config.
+	issuerPATH = "/dex"
+	// Dex callback path. Needs to be match the dex staticClients.redirectURIs helm config.
+	callbackPATH = "/callback"
+	// Kuberpult login path.
+	loginPATH = "/login"
+	// Dex OAUTH token name.
+	dexOAUTHTokenName = "kuberpult.oauth"
+	// Default value for the number of days the token is valid for.
+	expirationDays = 1
+)
+
+// NewDexAppClient a Dex Client.
+func NewDexAppClient(clientID, clientSecret, baseURL string, scopes []string) (*DexAppClient, error) {
+	a := DexAppClient{
+		ClientID:     clientID,
+		ClientSecret: clientSecret,
+		Scopes:       scopes,
+		BaseURL:      baseURL,
+		RedirectURI:  baseURL + callbackPATH,
+		IssuerURL:    baseURL + issuerPATH,
+	}
+	transport := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+	}
+	a.Client = &http.Client{
+		Transport: transport,
+	}
+
+	// Creates a transport layer to map all requests to dex internally
+	dexURL, _ := url.Parse(dexServiceURL)
+	a.Client.Transport = DexRewriteURLRoundTripper{
+		DexURL: dexURL,
+		T:      a.Client.Transport,
+	}
+
+	// Register Dex handlers.
+	a.registerDexHandlers()
+	return &a, nil
+}
+
+// DexRewriteURLRoundTripper creates a new DexRewriteURLRoundTripper.
+// The round tripper is configured to avoid exposing the dex server via a virtual service. Since Kuberpult and dex
+// are running on the same cluster, a reverse proxy is configured to redirect all dex calls internally.
+type DexRewriteURLRoundTripper struct {
+	DexURL *url.URL
+	T      http.RoundTripper
+}
+
+func (s DexRewriteURLRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	r.URL.Host = s.DexURL.Host
+	r.URL.Scheme = s.DexURL.Scheme
+	r.Host = s.DexURL.Host
+	return s.T.RoundTrip(r)
+}
+
+// Registers dex handlers for login
+func (a *DexAppClient) registerDexHandlers() {
+	// Handles the login callback to redirect to dex page.
+	http.HandleFunc(issuerPATH+"/", NewDexReverseProxy(dexServiceURL))
+	// Handles the login callback to redirect to dex page.
+	http.HandleFunc(loginPATH, a.handleDexLogin)
+	// Call back to the current app once the login is finished
+	http.HandleFunc(callbackPATH, a.handleCallback)
+}
+
+// NewDexReverseProxy returns a reverse proxy to the Dex server.
+func NewDexReverseProxy(serverAddr string) func(writer http.ResponseWriter, request *http.Request) {
+	target, err := url.Parse(serverAddr)
+	if err != nil {
+		logger.FromContext(context.Background()).Warn(fmt.Sprintf("Could not parse server URL with error: %s", err))
```

Comment:
we continue here despite the error?

Reply:
Added a logger with `Error()` and `return nil` statement.